lnwallet/wallet.go

```diff
@@ -474,41 +505,82 @@ func (l *LightningWallet) handleFundingReserveRequest(req *InitFundingReserveMsg
 		// Coin selection is done on the basis of sat/kw, so we'll use
 		// the fee rate passed in to perform coin selection.
 		var err error
-		selected, err = l.selectCoinsAndChange(
-			req.FundingFeePerKw, req.LocalFundingAmt, req.MinConfs,
-			req.SubtractFees,
+		fundingReq := &chanfunding.Request{
+			RemoteAmt:    req.RemoteFundingAmt,
+			LocalAmt:     req.LocalFundingAmt,
+			MinConfs:     req.MinConfs,
+			SubtractFees: req.SubtractFees,
+			FeeRate:      req.FundingFeePerKw,
+			ChangeAddr: func() (btcutil.Address, error) {
+				return l.NewAddress(WitnessPubKey, true)
+			},
+		}
+		fundingIntent, err = req.ChanFunder.ProvisionChannel(
+			fundingReq,
 		)
 		if err != nil {
 			req.err <- err
 			req.resp <- nil
 			return
 		}
 
-		localFundingAmt = selected.fundingAmt
+		localFundingAmt = fundingIntent.LocalFundingAmt()
+		remoteFundingAmt = fundingIntent.RemoteFundingAmt()
+
+	} else {
+		l.intentMtx.Lock()
+		// Alternatively, if we've just received an inbound funding
+		// request that we have a regisstered him intent to, then we'll
+		// obtain the backing intent now. In this case, we're doing a
+		// special funding workflow that allows more advanced
+		// constructions such as channel factories to be instantiated.
+		if shimIntent, ok := l.fundingIntents[req.PendingChanID]; ok {
+			fundingIntent = shimIntent
+		}
+		l.intentMtx.Unlock()
 	}
 
 	// The total channel capacity will be the size of the funding output we
 	// created plus the remote contribution.
-	capacity := localFundingAmt + req.RemoteFundingAmt
+	capacity := localFundingAmt + remoteFundingAmt
 
 	id := atomic.AddUint64(&l.nextFundingID, 1)
 	reservation, err := NewChannelReservation(
 		capacity, localFundingAmt, req.CommitFeePerKw, l, id,
 		req.PushMSat, l.Cfg.NetParams.GenesisHash, req.Flags,
-		req.Tweakless,
+		req.Tweakless, req.ChanFunder, req.PendingChanID,
 	)
 	if err != nil {
-		selected.unlockCoins()
+		if fundingIntent != nil {
+			fundingIntent.Cancel()
```

Comment:
Shouldn't we also do these for the other steps, e.g., `handleContributionMsg`?

Reply:
To free up the coins? If a reservation is failed, and doesn't proceed, then they'll be reclaimed via the zombie reservation sweeper within the `fundingManager`.

Reply:
Following that logic, why bother doing it here then? Seems inconsistent.

Reply:
It's done here optimistically to immediately clean up any coins for a funding request that hasn't even advanced passed the reservation phase. `failFundingFlow` already does this (in response to another error, or an invalid funding workflow message by the remote peer) since it calls `Cancel` on the reservation object, which will cancel and remove the intent.